### PR TITLE
Change report output format and make system tests run the maven plugin

### PR DESCRIPTION
**Pull request by spkrka**

### CODE REVIEW

**Comment by pettermahlen** [comment]
Some questions/comments:

1. I think we should eventually move away from doing a lot of work in the Mojo to making that just something that translates between the Maven world and 'ours'.
1. Does this change lead to the test artifacts being deployed as well as the 'real' ones? That would not be great.
1. I think it's not a great pattern to compare the exact expected outputs of the tests, because that makes the system tests brittle due to very tight coupling to the output format. A better way is specifying at a higher level that 'conflict X is reported', and then have some code that knows how to find conflict X in the output. That way, when the output changes, all you need to do is update that code, and all the tests would be fixed. With the current solution, every test file needs to change each time the output format changes - and the main problem with that is probably a mental lock-in on the current output format.
1. Finally, I think we should make JSON reporting and text output as before orthogonal. Either always output the text result, or make both of them togglable via configuration parameters.

So, in summary, I'm -1 on this PR as it currently stands. I think we should fix the last two items above before merging.

**Comment by spkrka** [comment]
# Agree
 # Probably. We should definitely fix that first.
 # I don't think this is a problem in practice. I made a script to copy actual to expected so we can easily verify the diff in git. I think this is a reasonable approach for system tests.
 # Perhaps. I think we should output everything to the report, but only output a small set for system out.

**Comment by spkrka** [comment]
I think I fixed 2. I still disagree on 3. It appears we are using a lot of memory when running on real artifacts which is why travis is sad.
4 can be done in a later commit.

**Comment by pettermahlen** [comment]
I feel strongly about 3, so we should probably try to discuss in real life so we can agree. It's a question of making tests that make it easier to make changes, not harder, and exact matching on outputs ties the tests very closely to the output format, making it a pain to change it. But let's discuss face-to-face instead.

Of course, the point of 4 is that I think it's quite likely people using the plugin won't be experts in how it works. That means there's a very big difference in usability of something that tells you in your (CI or your own) console what the problem is compared to something that tells you only in some more-or-less obscure report file. But we don't disagree on that, and if you prefer doing that in a later commit, that's OK.

**Comment by pettermahlen** [comment]
And of course - one more option, as long as you think that's OK, is for me to do the work of changing from exact matches to something else.

File: maven-plugin/src/main/java/com/spotify/missinglink/maven/CheckMojo.java [hunk 1]
```diff
@@ -550,56 +427,16 @@ private static long asMillis(Stopwatch stopwatch) {
     return stopwatch.elapsed(TimeUnit.MILLISECONDS);
   }
 
-  private void outputConflicts(Collection<Conflict> conflicts) {
+  private void outputConflicts(Collection<Conflict> conflicts, CheckReport checkReport) {
```
Comment (mattnworb) [inline maven-plugin/src/main/java/com/spotify/missinglink/maven/CheckMojo.java]: feels awkward to me to pass in the CheckReport as a parameter if this method is only called from one place, seems like the signature should just be `CheckReport outputConflicts(conflicts)`

File: maven-plugin/pom.xml [hunk 2]
```diff
@@ -21,6 +21,21 @@
...
     </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-core</artifactId>
+      <version>2.5.1</version>
+    </dependency>
+    <dependency>
+      <groupId>com.fasterxml.jackson.core</groupId>
+      <artifactId>jackson-annotations</artifactId>
+      <version>2.5.0</version>
```
Comment (mattnworb) [inline maven-plugin/pom.xml]: should this be `2.5.1` also?

**Comment by mattnworb** [comment]
I would prefer if the maven plugin continued to log the conflicts to `getLog()`, in addition to logging to a json file. I could be misunderstanding, but if the developer (not automated) workflow becomes a) run the plugin b) read a json file then we might be making it harder for users to see what problems there are.

If the standard out output is too verbose, then users can quiet that with `mvn -q` and/or we could add a parameter to disable that.

**Comment by caesar-ralf** [comment]
closing 8 years old PR. Feel free to re-open if you feel it's something that still could be useful
